services/offline/src/lib/online-status-message.tsx

```diff
@@ -1,39 +1,64 @@
 import React, { ReactElement, ReactNode, useContext, useState } from 'react'
-import { OnlineStatusMessageContextAPI } from '../types'
 
-const defaultApi: OnlineStatusMessageContextAPI = {
-    onlineStatusMessage: undefined,
-    setOnlineStatusMessage: () => undefined,
-}
-
-const OnlineStatusMessageContext =
-    React.createContext<OnlineStatusMessageContextAPI>(defaultApi)
+type SetOnlineStatusMessage = (message: ReactNode) => void
 
-export const useOnlineStatusMessage = (): OnlineStatusMessageContextAPI => {
-    const { onlineStatusMessage, setOnlineStatusMessage } =
-        useContext<OnlineStatusMessageContextAPI>(OnlineStatusMessageContext)
-
-    return {
-        onlineStatusMessage,
-        setOnlineStatusMessage,
-    }
-}
+// 'get' and 'set' contexts are separated so 'setter' consumers that don't
+// actually need the value don't have to rerender when the value changes:
+const OnlineStatusMessageValueContext = React.createContext<ReactNode>(null)
+const SetOnlineStatusMessageContext =
+    React.createContext<SetOnlineStatusMessage>(() => undefined)
 
 export const OnlineStatusMessageProvider = ({
     children,
 }: {
     children: ReactNode
 }): ReactElement => {
-    const [onlineStatusMessage, setOnlineStatusMessage] = useState<ReactNode>()
+    const [onlineStatusMessage, setOnlineStatusMessage] =
+        useState<ReactNode>(null) // note: not undefined
 
     return (
-        <OnlineStatusMessageContext.Provider
-            value={{
-                onlineStatusMessage,
-                setOnlineStatusMessage,
-            }}
-        >
-            {children}
-        </OnlineStatusMessageContext.Provider>
+        <OnlineStatusMessageValueContext.Provider value={onlineStatusMessage}>
+            <SetOnlineStatusMessageContext.Provider
+                value={setOnlineStatusMessage}
+            >
+                {children}
+            </SetOnlineStatusMessageContext.Provider>
+        </OnlineStatusMessageValueContext.Provider>
     )
 }
+
+export const useOnlineStatusMessageValue = () => {
+    const onlineStatusMessage = useContext(OnlineStatusMessageValueContext)
+
+    // note: value is initialized to `null` in provider, not undefined
+    if (onlineStatusMessage === undefined) {
+        throw new Error(
+            'useOnlineStatusMessageValue must be used within an OnlineStatusMessageProvider'
+        )
+    }
```

Comment:
This could cause issues if you manually set the status to ```undefined````. Subsequent calls to this would then throw.

So if we were to do this, I think we should prevent ```undefined``` being passed to `setOnlineStatusMessage`. However, if we just want to check that it's not the default, we could initialize it to a Symbol.

But this is all rendered in the app-shell, right? So this might just create more complexity than it's worth. And calling `setOnlineStatusMessage` without a context is fine, since it's just a noop. 

So in my opinion I think we could just remove these checks, which would work-around the "potential breaking change". 

Initializing with `null` or `undefined` then doesn't matter, and was never part of the API contract. And no functionality would change. The only problem would be these checks if we kept it like this (mostly due to being able to set it as `undefined` through `setOnlineStatusMessage`.

Reply:
If we really wanted to fail early, I think it would look something like this:

```ts
const defaultOnlineStatusMessage = Symbol('defaultStatusMessage')
const OnlineStatusMessageValueContext = React.createContext<ReactNode | typeof defaultOnlineStatusMessage >(defaultOnlineStatusMessage )


export const useOnlineStatusMessageValue = () => {
    const onlineStatusMessage = useContext(OnlineStatusMessageValueContext)

    // note: value is initialized to `null` in provider, not undefined
    if (onlineStatusMessage === defaultOnlineStatusMessage) {
        throw new Error(
            'useOnlineStatusMessageValue must be used within an OnlineStatusMessageProvider'
        )
    }
// this should be inferred to be a ReactNode
return onlineStatusMessage
```

`useSetOnlineStatusMessage`, would be similiar but extracting the "noop" default function to variable and check for that equality.

Reply:
That's a solid solution, but I just went with removing the undefined checks to keep it simple -- I think you're right that the app adapter kinda guarantees the provider will be there